### ISSUE TITLE
CA2000 False positive if returning result of method which may dispose the object

### ISSUE THREAD

**Issue by RobSiklos** [dotnet/roslyn-analyzers#2637]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

2.9.4-beta2.19326.1

#### Diagnostic ID

[CA2000: Dispose objects before losing scope](https://docs.microsoft.com/en-ca/visualstudio/code-quality/ca2000-dispose-objects-before-losing-scope)

#### Repro steps

1. Analyze the code below:

```csharp
public static class CA2000Issue
{
    private sealed class Thing : IDisposable
    {
        public void Dispose() { }
    }

    public static IDisposable DoSomething()
    {
        // CA2000 reported here - false positive.
        return GetThing();
    }

    private static IDisposable GetThing()
    {
        var thing = new Thing();

        // If you comment out this block, the warning does not occur.
        if (thing.GetHashCode() == 0)
        {
            thing.Dispose();
            return null;
        }

        return thing;
    }
}
```

#### Expected behavior
No CA2000 violation on `return GetThing()`.

#### Actual behavior
CA2000 violation on `return GetThing()`. However, if you comment out the `if` block in `GetThing()`, the warning goes away.

**Comment by RobSiklos** [comment]
Verified fixed in 2.9.4-beta2.19327.2

Thanks!
